### PR TITLE
feat/improve-presentation-exchange-support-in-oid4vp-rs

### PR DESCRIPTION
**Pull request by Ryanmtate**
add implementation methods for Presentation Definition.

WIP: Need to continue work for Presentation Submission and the rest of the structs used in the presentation exchange flow.

Many thanks to @todd-spruceid for pair programming on this.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -18,14 +18,15 @@ async-trait = "0.1.73"
...
+jsonpath_lib = "0.3.0"
 p256 = { version = "0.13.2", features = ["jwk"], optional = true }
 reqwest = { version = "0.12.5", features = ["rustls-tls"], optional = true }
 serde = "1.0.188"
 serde_cbor = "0.11.2"
 serde_json = "1.0.107"
 serde_qs = "0.12.0"
 serde_urlencoded = "0.7.1"
-ssi = "0.7.0"
+ssi = "0.7"
```
Comment (Ryanmtate) [inline Cargo.toml:31]: Waiting to bump the version until after the presentation exchange improvements.

File: src/json_schema_validation.rs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            && self.required == other.required
+            && self.properties == other.properties
+            && self.items == other.items
+    }
+}
+
+impl Eq for SchemaValidator {}
+
+impl SchemaValidator {
+    pub fn validate(&self, value: &Value) -> Result<(), String> {
```
Comment (Ryanmtate) [inline src/json_schema_validation.rs]: Replace with `anyhow::Error` type.

File: tests/e2e.rs [hunk 3]
```diff
@@ -16,18 +16,27 @@ mod jwt_vc;
...
     let (wallet, verifier) = jwt_vc::wallet_verifier().await;
 
+    // let
+
+    // let presentation_definition = PresentationDefinition::new(uuid::Uuid::new_v4())
+    //     .with_input_descriptors(input_descriptors)
+    //     .with_name(name)
+    //     .with_purpose(purpose)
+    //     .with_format(format)
+
```
Comment (Ryanmtate) [inline tests/e2e.rs]: TODO: Complete e2e test case updates using the new interface. Work to be completed in #8

File: Cargo.toml (hunk 1 above)
Comment (Ryanmtate) [reply Cargo.toml:31]: To be completed in #8

File: src/json_schema_validation.rs [hunk 4]
```diff
@@ -0,0 +1,204 @@
+use anyhow::{bail, Context, Result};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::collections::HashMap;
+
+// TODO: Consider using `Value` type from `serde_json`
```
Comment (cobward) [inline src/json_schema_validation.rs]: There's not an easy way to reuse that here, so I think it's fine to define this new enum

File: src/presentation_exchange.rs [hunk 5]
```diff
@@ -1,80 +1,500 @@
...
+/// Presentation Definitions are composed of inputs, which describe the forms and details of the
+/// proofs they require, and optional sets of selection rules, to allow [Holder](https://identity.foundation/presentation-exchange/spec/v2.0.0/#term:holder)s flexibility
+/// in cases where different types of proofs may satisfy an input requirement.
+///
+/// For more information, see: [https://identity.foundation/presentation-exchange/spec/v2.0.0/#presentation-definition](https://identity.foundation/presentation-exchange/spec/v2.0.0/#presentation-definition)
+#[derive(Clone, Default, Debug, Serialize, Deserialize, PartialEq, Eq)]
 pub struct PresentationDefinition {
-    pub id: String, // Uuid,
-    pub input_descriptors: Vec<InputDescriptor>,
+    id: uuid::Uuid, // TODO: The specification allows for non-uuid types, should we revert to using String type?
```
Comment (cobward) [inline src/presentation_exchange.rs]: Yep, best to keep it as a String.

File: src/json_schema_validation.rs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+                bail!(
+                    "String length {} is greater than maximum {}",
+                    s.len(),
+                    max_length
+                );
+            }
+        }
+
+        if let Some(pattern) = &self.pattern {
+            // Note: In a real implementation, you'd use a regex library here
```
Comment (Ryanmtate) [inline src/json_schema_validation.rs]: TODO: Add `regex` library for pattern matching

File: src/json_schema_validation.rs [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+            && self.maximum == other.maximum
+            && self.required == other.required
+            && self.properties == other.properties
+            && self.items == other.items
+    }
+}
+
+impl Eq for SchemaValidator {}
+
+impl SchemaValidator {
```
Comment (cobward) [inline src/json_schema_validation.rs]: Could you add a constructor?

File: src/json_schema_validation.rs [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    pub fn validate(&self, value: &Value) -> Result<()> {
```
Comment (cobward) [inline src/json_schema_validation.rs]: Could you add some unit tests for this?

File: src/presentation_exchange.rs [hunk 9]
```diff
@@ -1,80 +1,584 @@
...
+    /// Return the name of the constraints field.
+    pub fn name(&self) -> Option<&String> {
+        self.name.as_ref()
+    }
+
+    /// Set the filter of the constraints field.
+    ///
+    /// If present its value MUST be a JSON Schema descriptor used to filter against
+    /// the values returned from evaluation of the JSONPath string expressions in the path array.
+    pub fn set_filter(mut self, filter: SchemaValidator) -> Self {
```
Comment (cobward) [inline src/presentation_exchange.rs]: I might be wrong, but it appears that `SchemaValidator` is not publically accessible outside of this crate.

File: tests/e2e.rs [hunk 10]
```diff
@@ -68,9 +69,9 @@ async fn w3c_vc_did_client_direct_post() {
             "definition_id": "0b4dd017-efa6-4a05-a269-9790fa3c22c2",
             "descriptor_map": [
                 {
-                    "id": "vc",
+                    "id": "064255a8-a0fa-4108-9ded-429f83003350",
                     "path": "$",
-                    "format": "jwt_vc_json"
+                    "format": "jwt_vp"
```
Comment (cobward) [inline tests/e2e.rs:74]: This is technically not correct, currently the submission is for a `jwt_vc_json`.

File: src/presentation_exchange.rs [hunk 11]
```diff
@@ -1,80 +1,584 @@
...
+///
+/// All Input Descriptors MUST be satisfied, unless otherwise specified by a
+/// [Feature](https://identity.foundation/presentation-exchange/spec/v2.0.0/#term:feature).
+///
+/// See: [https://identity.foundation/presentation-exchange/spec/v2.0.0/#input-descriptor-object](https://identity.foundation/presentation-exchange/spec/v2.0.0/#input-descriptor-object)
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
 pub struct InputDescriptor {
-    pub id: String,
+    id: String,
+    constraints: Constraints,
```
Comment (cobward) [inline src/presentation_exchange.rs:306]: Is there any reason we shouldn't add a `#[serde(default)]` attribute on `constraints` for some backwards-compatibility?

**Review by cobward — CHANGES_REQUESTED** [review]
Looking really good, a couple more points of feedback. Once addressed and the CI is passing, then we can look to merge!

File: src/json_schema_validation.rs (hunk 8 above)
Comment (Ryanmtate) [reply src/json_schema_validation.rs]: Commit [d1884](https://github.com/spruceid/oid4vp-rs/pull/9/commits/d188434954c571953cd572c547e070e8e5524d51) adds unit tests.

File: src/json_schema_validation.rs (hunk 7 above)
Comment (Ryanmtate) [reply src/json_schema_validation.rs]: Constructor now takes a `SchemaType` to set the type of the validation.

File: tests/e2e.rs (hunk 10 above)
Comment (Ryanmtate) [reply tests/e2e.rs:74]: This will be fixed in #8.

File: src/presentation_exchange.rs (hunk 9 above)
Comment (Ryanmtate) [reply src/presentation_exchange.rs]: Made the module public.

File: src/presentation_exchange.rs (hunk 11 above)
Comment (Ryanmtate) [reply src/presentation_exchange.rs:306]: [89db3b2](https://github.com/spruceid/oid4vp-rs/pull/9/commits/89db3b238a56dc5bd2ae43cee9cb41f9fd526817) adds this tag. This should help with backwards compatibility for schema types that don't have a `constraints` field. Although, they should, according to the spec. Might need to update the test cases to include, but at least this should resolve any compatibility-related issues. Possibly the problem with the CI.

**Comment by Ryanmtate** [comment]
Todd helped point out that I need to add the submodules for the presentation-exchange. Will re-run tests now that I have the tests locally.

**Review by cobward — CHANGES_REQUESTED** [review]
The CI is failing due to the presentation exchange tests.

**Comment by Ryanmtate** [comment]
[d18f56d](https://github.com/spruceid/oid4vp-rs/pull/9/commits/d18f56d265ae4c7e710de5d8eb0469572df2bd1f) should fix the CI test failures.

File: src/presentation_exchange.rs [hunk 12]
```diff
@@ -447,9 +466,8 @@ pub struct ConstraintsField {
     purpose: Option<String>,
     #[serde(skip_serializing_if = "Option::is_none")]
     name: Option<String>,
-    // TODO: JSONSchema validation at deserialization time
     #[serde(skip_serializing_if = "Option::is_none")]
-    filter: Option<SchemaValidator>,
+    filter: Option<serde_json::Value>,
```
Comment (Ryanmtate) [inline src/presentation_exchange.rs:470]: Replacing `SchemaValidator` to use a generic `Value` type and using `jsonschema` to perform the validation.

SchemaValidator provides helpful methods to construct a filter object and will perform some validation, but it is not as feature rich as `jsonschema`.

Consider removing `SchemaValidator` from the code, unless there is a reason to keep it.

Comment (cobward) [reply src/presentation_exchange.rs:470]: I would be in favour of removing it, I think we eventually need a more featureful builder than what it currently is anyway.

Comment (Ryanmtate) [reply src/presentation_exchange.rs:470]: I'll wire up the schema validation (using jsonschema) in #8 and remove the schema validator from this PR.

**Review by cobward — APPROVED** [review]

**Pull request merged**